src/dbnode/persist/fs/seek_manager.go

```diff
@@ -496,6 +500,51 @@ func (m *seekerManager) updateOpenLeaseHotSwapSeekers(
 	return wg, updateOpenLeaseResult, nil
 }
 
+func (m *seekerManager) RelinquishShard(
+	namespace ident.ID,
+	shard uint32,
+) error {
+	if !m.namespace.Equal(namespace) {
+		return fmt.Errorf("seekerManager's namespace %q != provided namespace %q",
+			m.namespace.String(), namespace,
+		)
+	}
+
+	byTime := m.seekersByTime(shard)
+
+	byTime.Lock()
+	defer byTime.Unlock()
+
+	for _, seekers := range byTime.seekers {
```

Comment:
So this algorithm doesn't quite work because some of the seekers may still be borrowed by `retriever.go` so you can't just return them here since they're still "outstanding".

I think you'll need to do something similar to what I did in `UpdateOpenLease` where you do something to mark the seekers as inactive or no longer valid, and then when they get returned they checked if they are the last seeker to get returned and close themselves (see the logic in `returnSeekerWithLock` for example)

Reply:
I tried a new algorithm